### ISSUE TITLE
Unable to Dynamically Download Drivers over CMG

### ISSUE THREAD

**Issue by Jhanno** [MSEndpointMgr/ConfigMgr#279]
Hi all,

I'm on the latest version of the Driver Automation Tool (6.4.9) and using the latest  Invoke-CMApplyDriverPackage.ps1 (4.0.5).

I'm able to download drivers successfully while on the **Intranet,** but havent had any success over the **Internet** (Cloud Management Gateway).

Here's the logs that I see for ApplyDriverPackage.log while on the Internet via CMG:

[ApplyDriverPackage]: Apply Driver Package process initiated
 - Apply driver package deployment type: PreCache
 - Apply driver package operational mode: Production
[PrerequisiteChecker]: Starting environment prerequisite checker
 - Supported computer platform detected, script execution allowed to continue
 - Computer manufacturer determined as: Dell
 - Computer model determined as: XPS 15 9570
 - Computer SystemSKU determined as: 087C
 - Computer Fallback SystemSKU determined as: 087C
 - Computer model detection was successful
 - Computer SystemSKU detection was successful
 - Computer details successfully verified
 - Determined primary computer detection method: SystemSKU
[PrerequisiteChecker]: Completed environment prerequisite checker
[AdminService]: Starting AdminService endpoint phase
 - Attempting to determine AdminService endpoint type based on current active Management Point candidates and from ClientInfo class
 - Determined AdminService endpoint type as: External
 - Successfully read service account user name from TS environment variable 'MDMUserName': S*C*_*S
 - Successfully read service account password from TS environment variable 'MDMPassword': *********
 - Successfully read external endpoint address for AdminService through CMG from TS environment variable 'MDMExternalEndpoint': HTTPS://CMG.COM/CCM_Proxy_ServerAuth/RandomDigits/AdminService
 - Successfully read client identification for AdminService through CMG from TS environment variable 'MDMClientID': ClientID
 - Successfully read client identification for AdminService through CMG from TS environment variable 'MDMTenantName': organization.onmicrosoft.com
 - Setting 'AdminServiceURL' variable to: HTTPS://CMG.COM/CCM_Proxy_ServerAuth/RandomDigits/AdminService/wmi
 - Attempting to locate PSIntuneAuth module
 - Authentication module detected, checking for latest version
 - Attempting to retrieve authentication token using native client with ID: ClientID

The task sequence essentially just gets stuck at this point. Any ideas?

**Comment by GbCaa** [comment]
I'm having the same problem. If you check the SMSTS.log, you'll see an error message was captured by it: WARNING: Failure to acquire access token. Response with access token was null

**Comment by tonycontorno** [comment]
Hey there, I'm experiencing the same problem here.  Could you elaborate on the URL you used for your CMG App Server?  Did you go down to /AdminService/wmi or was there another URL you were using?

**Comment by GbCaa** [comment]
> Hey there, I'm experiencing the same problem here. Could you elaborate on the URL you used for your CMG App Server? Did you go down to /AdminService/wmi or was there another URL you were using?

In Azure "App Registrations", find the one that points to your SCCM server... in the overview, the "Application ID URI" is the URL I was using for the -Resource parameter.

**Comment by tonycontorno** [comment]
Ah, thanks.  That could explain part of our problem.  I don't have that field filled out at all.  I suspect something was missed...

**Comment by gitsyscu** [comment]
> > Hey there, I'm experiencing the same problem here. Could you elaborate on the URL you used for your CMG App Server? Did you go down to /AdminService/wmi or was there another URL you were using?
> 
> In Azure "App Registrations", find the one that points to your SCCM server for Cloud Management Gateway... in the overview, the "Application ID URI" is the URL I was using for the -Resource parameter.

Forgive my ignorance, but where are you seeing the -Resource parameter?

**Comment by Jhanno** [comment]
> > Hey there, I'm experiencing the same problem here. Could you elaborate on the URL you used for your CMG App Server? Did you go down to /AdminService/wmi or was there another URL you were using?
> 
> In Azure "App Registrations", find the one that points to your SCCM server for Cloud Management Gateway... in the overview, the "Application ID URI" is the URL I was using for the -Resource parameter.

I also have dont "Application ID URI" filled out.... was that in the documentation? I cant seem to find that step. How did you set this up?

**Comment by GbCaa** [comment]
> > > Hey there, I'm experiencing the same problem here. Could you elaborate on the URL you used for your CMG App Server? Did you go down to /AdminService/wmi or was there another URL you were using?
> > 
> > 
> > In Azure "App Registrations", find the one that points to your SCCM server for Cloud Management Gateway... in the overview, the "Application ID URI" is the URL I was using for the -Resource parameter.
> 
> I also have dont "Application ID URI" filled out.... was that in the documentation? I cant seem to find that step. How did you set this up?

It's not in the documentation. I'm betting the author assumed you followed their documentation for setting up a Cloud Management Gateway in which case the -Resource would be the same name. On line 812 of the Invoke-CMApplyDriverPackage, I changed my resource to the name I used when I created the Application ID URI:

`			$Script:AuthToken = Get-MSIntuneAuthToken -TenantName $TenantName -ClientID $ClientID -Credential $Credential -Resource "https://cmg/ConfigMgrService" -RedirectUri "https://login.microsoftonline.com/common/oauth2/nativeclient" -ErrorAction Stop
`

**Comment by vbateman** [comment]
So by default your CMG app will be https://ConfigMgrService (unless you changed - which I didn't) And line 812 (actually line 814 in the latest script) will refer to that. But I'm having no luck getting past this in the log: "Attempting to retrieve authentication token using native client with ID: ClientID", it eventually times out.

Anyone else got this working using all the defaults when creating the Web App??

**Comment by TacII** [comment]
> So by default your CMG app will be https://ConfigMgrService (unless you changed - which I didn't) And line 812 (actually line 814 in the latest script) will refer to that. But I'm having no luck getting past this in the log: "Attempting to retrieve authentication token using native client with ID: ClientID", it eventually times out.
> 
> Anyone else got this working using all the defaults when creating the Web App??

Yes, I had the same problem. Make sure that the account you're using to access the Web App doesn't have MFA enabled. You should be able to verify that everything works by running this command. Replace your Tenant Name and ClientID of course.

Get-MSIntuneAuthToken -TenantName "yourtenant.onmicrosoft.com" -ClientID "12345678-1234-1234-1234-123412341234" -RedirectUri "https://login.microsoftonline.com/common/oauth2/nativeclient" -Resource "https://ConfigMgrService" -verbose

You will get a popup that asks for your service account credentials. Be sure to enter it with the full domainname.

**Comment by JackS369** [comment]
Hi @TacII, when i run the above command on the client machine, I'm getting the error of "Warning: Failure to acquire access token. Response with access token was null".   Any advice on this?       @GbCaa @Jhanno   is It working for you now?

**Comment by GbCaa** [comment]
> Hi @TacII, when i run the above command on the client machine, I'm getting the error of "Warning: Failure to acquire access token. Response with access token was null". Any advice on this? @GbCaa @Jhanno is It working for you now?

Yes and with the latest script by added in to the list of Dynamic Task Sequence Variables MDMApplicationIDURI=https://cmg/ConfigMgrService (my CMG as found in "App Registrations")

If you're still stuck, check the Sign-Ins & Audit logs of your service account in Azure AD. Hopefully it will give you a clue.

**Comment by MarkusMory** [comment]
Good morning,

@gregor79 I got exactly the same problem at one of my customers. If I run `
Get-MSIntuneAuthToken -TenantName $TenantName -ClientID $ClientID -Credential $Credential -Resource $ApplicationIDURI -RedirectUri "https://login.microsoftonline.com/common/oauth2/nativeclient" -Verbose`
the command will fail because of the basic (legacy) authentication. If I run the same command without the -Credential parameter it will use the modern authentication. This way the authentication works correctly.

Enabling basic authentication seems to be a workaround but if I remeber it correctly MS is going to block basic authentication automatically 
https://azurescene.com/2021/01/18/an-update-on-legacy-authentication/

**Comment by gregor79** [comment]
@MarkusMory 
Thanks for your comment. It still doesn't work for me though when I remove -Credential parameter. As script runs in local system context when acquiring access token it is presented with the modern auth prompt. How do I pass on credentials in a script?

VERBOSE: AzureAD module detected, checking for latest version
VERBOSE: Latest version for AzureAD module was detected, continue to aquire authentication token
VERBOSE: Currently running in parameter set context: AuthPrompt
VERBOSE: Attempting to acquire access token using user delegation

**Comment by MarkusMory** [comment]
@gregor79 
Sorry for the late response. I think you can't use credentials here. There are two other ways for authenticating against an Azure App via Script. You can use a certificate or a client secret (https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal#authentication-two-options). 

I will test the secret key method as soon as I have time for it.

**Comment by bbrownchg** [comment]
> @gregor79
> Sorry for the late response. I think you can't use credentials here. There are two other ways for authenticating against an Azure App via Script. You can use a certificate or a client secret (https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal#authentication-two-options).
> 
> I will test the secret key method as soon as I have time for it.

I was able to modify the script to use a client secret to generate a token, which the logs show was successful. After that, though, I get a 401 unauthorized when trying to query for drivers. Trying to troubleshoot that right now.

**Comment by MarkusMory** [comment]
I guessed that this would be a problem. You can authenticate against the Azure app but not against the MECM Admin Service with the secret key. Maybe @maurice-daly  has an idea how to fix that.

**Comment by bbrownchg** [comment]
> I guessed that this would be a problem. You can authenticate against the Azure app but not against the MECM Admin Service with the secret key. Maybe @maurice-daly has an idea how to fix that.

Yea, that makes sense and is pretty much what I figured yesterday. I did go back to the drawing board though and figured out that if I removed the `-Resource $ApplicationIDURI` line from `Get-MSIntuneAuthToken` that I'm able to generate a token using the credentials, but I'm still running in to the 401 on the actual query. Waiting for my Azure AD logs to update to see what the issue is.

**Comment by bbrownchg** [comment]
Well I still haven't made any progress here. I figured out my application ID URI was wrong and fixed that, am able to generate a token while specifying that but I'm still getting a 401 unauthorized trying to query over the CMG. The same account can query on the internal network no problem. The confusing thing to me is the external query doesn't even show in the AdminService.log file. Probably gonna construct a Reddit post here to see if the people smarter than me can figure out what I'm doing wrong LOL.

**Comment by bbrownchg** [comment]
Found it in the logs, it seems like it doesn't like something with the tokens.

`ERROR: Security token validation exception with requesting URL https://mycmg.example.com/CCM_Proxy_ServerAuth/72057594037927941/AdminService/wmi/SMS_Package?$filter=contains(Name,"Drivers"). System.IdentityModel.Tokens.SecurityTokenValidationException: There is no appropriate token validator to validate the token with issuer https://sts.windows.net/my-tenant-id/ and audience https://myapplicationiduri~~   at Microsoft.ConfigurationManager.CloudBase.AuthorizationToken.TokenValidator.ValidateJsonToken(String token)~~   at Microsoft.ConfigurationManager.CloudBase.AuthorizationToken.TokenValidator.ValidateTokenEx(String token, String tokenHint)~~   at Microsoft.ConfigurationManager.BgbServerChannel.BgbServerReverseProxy.ValidateAuthorizationToken(String authorizationToken, EndpointClientAuthScheme clientAuthScheme, Uri requestUri, IToken& validatedToken, EndpointClientAuthScheme& validatedScheme)`

After doing some reading I noticed that this was generating a token against the v1 endpoint and not v2. To determine if this was the root cause, I updated the manifest on my app registrations to enable the v2 endpoint: `"accessTokenAcceptedVersion": 2,`

After doing that I'm able to generate a token on the v2 endpoint but still end up with the 401 unauthorized:

`ERROR: Security token validation exception with requesting URL https://mycmg.example.com/CCM_Proxy_ServerAuth/72057594037927941/AdminService/wmi/SMS_Package?$filter=contains(Name,"Drivers"). System.IdentityModel.Tokens.SecurityTokenValidationException: There is no appropriate token validator to validate the token with issuer https://login.microsoftonline.com/my-tenant-id/v2.0 and audience my-app-registration-id~~   at Microsoft.ConfigurationManager.CloudBase.AuthorizationToken.TokenValidator.ValidateJsonToken(String token)~~   at Microsoft.ConfigurationManager.CloudBase.AuthorizationToken.TokenValidator.ValidateTokenEx(String token, String tokenHint)~~   at Microsoft.ConfigurationManager.BgbServerChannel.BgbServerReverseProxy.ValidateAuthorizationToken(String authorizationToken, EndpointClientAuthScheme clientAuthScheme, Uri requestUri, IToken& validatedToken, EndpointClientAuthScheme& validatedScheme)	CMGService	6/9/2021 1:04:15 PM	31 (0x001F)`

At this point I'm not sure if I'm going down the wrong rabbit hole or if I'm on to something 🤣.

**Comment by bbrownchg** [comment]
Just in case anyone else ran in to the problem I did, here's how I fixed it:

I noticed some inconsistencies from when I set up the CMG, I had a couple of different server apps configured in Azure. I was trying to reconfigure them and wasn't able to sign in to Azure via the console and came across this article: [https://systemcenterdudes.com/failed-to-sign-in-to-azure-error-when-configuring-sccm-cloud-management-gateway/]. Pretty much, I blew away my Azure services and tenant in my console and reconnected and created new server/client apps. Everything is working now!

**Comment by MarkusMory** [comment]
Cool that you got it working. But are you using a client secret or the basic authentication with username/password?

**Comment by bbrownchg** [comment]
> Cool that you got it working. But are you using a client secret or the basic authentication with username/password?

Client secret won't work to auth against the AdminService as it requires ConfigMgr RBAC on the backend for all authentication. Until ConfigMgr is updated to understand something more than AD auth that'll remain the case. I'd have to imagine that MS has something in the works for this as they very much preach using service principals instead of traditional username/password authentication where possible.

I'm still using the default script which provides the MDM variables to set the credentials. Looking at the sign-in logs for the service account shows the client app as "Mobile Apps and Desktop clients" as it's using the ADAL library to generate the token. 

Now, to go on what @gregor79 was saying earlier, my service account does redirect to my ADFS instance for authentication where we do still allow username/password authentication. I'm not sure how to workaround that piece without modifying the ADFS config. An alternative would be to set up a service account using a different domain name that uses managed authentication instead of federated. That would depend on the individual setup though.
